core/src/main/java/io/questdb/cairo/map/OrderedMap.java

```diff
@@ -1043,10 +1043,7 @@ public void putTimestamp(long value) {
         public void putVarchar(Utf8Sequence value) {
             int byteCount = VarcharTypeDriver.getSingleMemValueByteCount(value);
             checkCapacity(byteCount);
-            // ASCII flag is best-effort, so there is a chance that identical ASCII-only strings
-            // have different values of the flag. This would break hash table look-ups, so we always
-            // set the flag to false.
-            VarcharTypeDriver.appendPlainValue(appendAddress, value, true);
+            VarcharTypeDriver.appendPlainValue(appendAddress, value, false);
```

Comment:
This may break group by and join correctness. At the moment, we don't assert ascii flag correctness in CSV and PGWire tests - probably the simplest way to do that would be to introduce `is_ascii` SQL function that would return `us.isAscii()`. This function may be also handy for us in future, when debugging perf/correctness issues for users.

Another, even more critical, problem is that ascii flag is currently best-effort and many utilities don't calculate it properly. A few examples: `CharSink#putUtf8()` implementations, `Utf8s#strCpy`.

Example of group by broken by this change:
```sql
create table x (i int, v varchar);
insert into x values (0, 'foobar');
insert into x values (1, 'foo');

select v, max(i)
from (
  select i, left(v, 3)::varchar as v from x where i = 0
  union all
  select i, v from x where i = 1
);
```
![Screenshot from 2024-04-15 09-32-03](https://github.com/questdb/questdb/assets/37772591/ddfc06d0-093e-4dda-936b-cc35db2c359e)

Reply:
this PR enforces ascii correctness. I will check the PGWire too! Everytime we assert SQL result we also assert ascii correctness. It may well be we don't have sufficient non-ascii tests for PGWire. I intend on adding them. CSV ascii is already assured and enforced.

Reply:
> this PR enforces ascii correctness

It adds some checks, but we don't have any specific tests aimed to verify the correctness. As the above message shows, there are scenarios where this change breaks group by/join. There may be even more of them, as it's hard to consider all SQL function combinations and other code paths. Are we planning to address them as known issues and fix them later?

Reply:
i will try to address all of them in this PR

Reply:
on this branch:

![image](https://github.com/questdb/questdb/assets/7276403/1b52c949-3fb9-470e-b6fd-f9dce6ff2f81)